arbnode/batch_poster.go

```diff
@@ -1069,6 +1070,121 @@ func (s *batchSegments) SetWaitingForValidation() {
+	}
+	// initially constructing the calldata using the old oldSequencerBatchPostMethodName method
+	// This will allow us to get the attestation quote on the hash of the data
+	args = append(args, seqNum)
+	args = append(args, l2MessageData)
+	args = append(args, new(big.Int).SetUint64(delayedMsg))
+	args = append(args, b.config().gasRefunder)
+	args = append(args, new(big.Int).SetUint64(uint64(prevMsgNum)))
+	args = append(args, new(big.Int).SetUint64(uint64(newMsgNum)))
+
+	// Later append the delay proof if needed for getting the attestion quote.
+	// If not, only append at the end of the calldata as done below.
+	calldata, err := method.Inputs.Pack(args...)
+	if err != nil {
+		return nil, err
+	}
+
+	attestationQuote, err := b.streamer.getAttestationQuote(calldata)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get attestation quote: %w", err)
+	}
+	// Construct the calldata with attestation quote
+	method, ok = b.seqInboxABI.Methods[newSequencerBatchPostMethodName]
+	if !ok {
+		return nil, errors.New("failed to find add batch method")
+	}
+	args = append(args, attestationQuote)
+
+	calldata, err = method.Inputs.Pack(args...)
+
+	if err != nil {
+		return nil, err
+	}
+
+	fullCalldata := append([]byte{}, method.ID...)
+	fullCalldata = append(fullCalldata, calldata...)
+	return fullCalldata, nil
+}
+
+func (b *BatchPoster) getCalldataForEspressoBlobBatch(
+	seqNum *big.Int,
+	prevMsgNum arbutil.MessageIndex,
+	newMsgNum arbutil.MessageIndex,
+	l2MessageData []byte,
+	delayedMsg uint64,
+) ([]byte, error) {
+	var args []any
+	method, ok := b.seqInboxABI.Methods[oldSequencerBatchPostWithBlobsMethodName]
+	if !ok {
+		return nil, errors.New("failed to find add batch method")
+	}
+	kzgBlobs, err := blobs.EncodeBlobs(l2MessageData)
+	_, blobHashes, err := blobs.ComputeCommitmentsAndHashes(kzgBlobs)
+	// initially constructing the calldata using the old SequencerBatchPostWithBlobsMethodName method
+	// This will allow us to get the attestation quote on the hash of the data
+	args = append(args, seqNum)
+	args = append(args, new(big.Int).SetUint64(delayedMsg))
+	args = append(args, b.config().gasRefunder)
+	args = append(args, new(big.Int).SetUint64(uint64(prevMsgNum)))
+	args = append(args, new(big.Int).SetUint64(uint64(newMsgNum)))
+
+	// Later append the delay proof if needed for getting the attestion quote.
+	// If not, only append at the end of the calldata as done below.
+	calldata, err := method.Inputs.Pack(args...)
+	if err != nil {
+		return nil, err
+	}
+	// bytes32Type, err := abi.NewType("bytes32", "", nil)
+	// if err != nil{
+	//   return nil, err
+	// }
+	bytes32ArrayType, err := abi.NewType("bytes32[]", "", nil)
+	if err != nil {
+		return nil, err
+	}
+	packedBlobHashes, err := abi.Arguments{
+		{Type: bytes32ArrayType},
+	}.Pack(blobHashes)
+	quoteData := append(calldata, packedBlobHashes...)
+	attestationQuote, err := b.streamer.getAttestationQuote(quoteData)
+	log.Info("Attestation Quote:", "quote", attestationQuote)
```

Comment:
I don't know how this is implemented on contract side. It looks weird for me to `pack` first and `append` data because it would be troublesome for the contract to decode. And is it even possible to be decoded back?

Reply:
Yeah I think my understanding of how abi.encode would pack the dynamic data was flawed. I left a reply to sneh's comment, does this line up with your understanding?